File: Yeti/Controllers/FeedVC/ArticleCell.m
//
//  ArticleCell.m
//  Yeti
//
//  Created by Nikhil Nigade on 17/05/20.
//  Copyright © 2020 Dezine Zync Studios. All rights reserved.
//

#import "ArticleCell.h"
#import "PrefsManager.h"
#import "FeedVC.h"

#import <DZKit/NSString+Extras.h>

#import "UIImage+Sizing.h"
#import "CheckWifi.h"
#import "NSString+ImageProxy.h"
#import "Paragraph.h"

#import <SDWebImage/UIImageView+WebCache.h>
#import <SDWebImage/UIImage+Transform.h>
#import <SDWebImage/SDWebImageError.h>

#if TARGET_OS_MACCATALYST

#import "AppDelegate.h"

// https://gist.github.com/steipete/9b279c94a35389c05bf5ea32336551ed
@implementation UIImage (ResourceProxyHack)

+ (UIImage *)_iconForResourceProxy:(id)proxy format:(int)format {
    
    // using this causes the app to use large amounts of memory.
    // so we simply return nil for now until Apple implements it
    // for catalyst 
    return nil;
}

@end

#endif

NSString *const kArticleCell = @"com.yeti.cell.article";

@interface ArticleCell () {
    
    BOOL _isShowingCover;
    
}

@property (nonatomic, strong) SDWebImageCombinedOperation *faviconTask;
@property (weak, nonatomic) IBOutlet NSLayoutConstraint *titleLabelWidthConstraint;
@property (nonatomic, assign) FeedVCType feedType;

@end

@implementation ArticleCell

+ (void)registerOnTableView:(UITableView *)tableView {
    
    if (tableView == nil) {
        return;
    }
    
    Class class = ArticleCell.class;
    NSBundle *bundle = [NSBundle bundleForClass:class];
    UINib *nib = [UINib nibWithNibName:NSStringFromClass(class) bundle:bundle];
    
    [tableView registerNib:nib forCellReuseIdentifier:kArticleCell];
    
}

- (void)awakeFromNib {
    [super awakeFromNib];
    // Initialization code
    
    self.coverImage.layer.cornerRadius = 3.f;
    self.coverImage.layer.cornerCurve = kCACornerCurveContinuous;
    
    self.titleLabel.textColor = UIColor.labelColor;
    self.summaryLabel.textColor = UIColor.secondaryLabelColor;
    
    self.authorLabel.textColor = UIColor.secondaryLabelColor;
    self.timeLabel.textColor = UIColor.secondaryLabelColor;
    
    self.selectedBackgroundView = [UIView new];
    
#if TARGET_OS_MACCATALYST
    self.selectedBackgroundView.backgroundColor = UIColor.secondarySystemBackgroundColor;
#else
    self.selectedBackgroundView.backgroundColor = [self.tintColor colorWithAlphaComponent:0.3f];
#endif
    
    [self resetUI];
    
}

- (void)prepareForReuse {
    
    [super prepareForReuse];
    
    [self resetUI];
    
}

- (void)resetUI {
    
    [self.coverImage il_cancelImageLoading];
    
    NSArray <UILabel *> * labels = @[self.titleLabel, self.summaryLabel, self.authorLabel, self.timeLabel];
    
    for (UILabel *label in labels) {
        
        label.text = nil;
        
    }
    
    self.coverImage.image = nil;
    
    self.coverImage.hidden = NO;
    
    self.markerView.image = nil;
    
    if (self.faviconTask) {
        [self.faviconTask cancel];
        self.faviconTask = nil;
    }
    
    self.semanticContentAttribute = UISemanticContentAttributeUnspecified;
    self.titleLabel.textAlignment = NSTextAlignmentLeft;
    self.authorLabel.textAlignment = NSTextAlignmentLeft;
    
}

- (void)tintColorDidChange {
    
    [super tintColorDidChange];
    
    [self updateMarkerView];
    
    self.selectedBackgroundView.backgroundColor = [SharedPrefs.tintColor colorWithAlphaComponent:0.3f];
    
}

#pragma mark - Configurations

- (BOOL)showImage {
    
    if ([SharedPrefs.imageBandwidth isEqualToString:ImageLoadingNever])
        return NO;
    
    else if([SharedPrefs.imageBandwidth isEqualToString:ImageLoadingOnlyWireless]) {
        return CheckWiFi();
    }
    
    return YES;
}

- (void)configure:(FeedItem *)article feedType:(NSInteger)feedType {
    
    if (article == nil) {
        _article = article;
        return;
    }
    
    if ([self showImage] == NO || SharedPrefs.articleCoverImages == NO) {
        
        self.coverImage.hidden = YES;
        
    }
    
    self.article = article;
    self.feedType = feedType;
    
    Feed *feed = [ArticlesManager.shared feedForID:self.article.feedID];
    
    [self configureTitle:feedType];
    
    NSString *coverImageURL = article.coverImage;
    
    if (coverImageURL == nil && article.content != nil && article.content.count > 0) {
        // find the first image
        Content *content = [article.content rz_reduce:^id(Content *prev, Content *current, NSUInteger idx, NSArray *array) {
            
            if (prev && [prev.type isEqualToString:@"image"]) {
                return prev;
            }
            
            return current;
        }];
        
        if (content != nil) {
            article.coverImage = content.url;
            coverImageURL = article.coverImage;
        }
    }
    
    [self configureSummary];
    
    [self configureAuthorWithFeedType:feedType feed:feed];
    
    BOOL isMicroBlogPost = NO;
    
    if ([([article articleTitle] ?: @"") isBlank] && article.content && article.content.count) {
        
        // find the first paragraph
        Content *content = [article.content rz_reduce:^id(Content *prev, Content *current, NSUInteger idx, NSArray *array) {
            
            if (prev && [prev.type isEqualToString:@"paragraph"]) {
                return prev;
            }
            
            return current;
            
        }];
        
        if (content) {
            
            isMicroBlogPost = YES;
            
            NSString * titleContent = [self.article textFromContent];
            
            self.titleLabel.text = titleContent;
            self.titleLabel.numberOfLines = MAX(3, SharedPrefs.previewLines);
            self.titleLabel.font = [UIFont preferredFontForTextStyle:UIFontTextStyleBody];
            
        }
        
    }
    
    if (isMicroBlogPost == NO) {
        self.titleLabel.font = [UIFont preferredFontForTextStyle:UIFontTextStyleHeadline];
    }
    
    self.titleLabel.accessibilityValue = [self.titleLabel.text stringByReplacingOccurrencesOfString:@" | " withString:@" by "];
    
    if (([Paragraph languageDirectionForText:article.articleTitle] == NSLocaleLanguageDirectionRightToLeft)
        || (article.summary && [Paragraph languageDirectionForText:article.summary] == NSLocaleLanguageDirectionRightToLeft)) {
        
        self.semanticContentAttribute = UISemanticContentAttributeForceRightToLeft;
        self.titleLabel.textAlignment = NSTextAlignmentRight;
        
        self.authorLabel.textAlignment = NSTextAlignmentRight;
        
    }
    
    [self updateMarkerView];
    
    _isShowingCover = NO;
    
    if ([self showImage]
#if !TARGET_OS_MACCATALYST
        && SharedPrefs.articleCoverImages == YES
#endif
        && article.coverImage != nil) {
        // user wants cover images shown
        _isShowingCover = YES;
        
        [self configureCoverImage:coverImageURL];
        
    }
    else {
        // do not show cover images
        self.coverImage.image = nil;
        self.coverImage.hidden = YES;
    }
    
    CGFloat width = self.bounds.size.width - 48.f;
        
    self.titleLabel.preferredMaxLayoutWidth = width - (_isShowingCover ? 92.f : 4.f); // 80 + 12
    self.titleLabelWidthConstraint.constant = self.titleLabel.preferredMaxLayoutWidth;
    
    self.summaryLabel.preferredMaxLayoutWidth = width;
    
    self.timeLabel.preferredMaxLayoutWidth = 92.f;
    
    self.authorLabel.preferredMaxLayoutWidth = self.titleLabel.preferredMaxLayoutWidth - 24.f - 12.f - self.timeLabel.preferredMaxLayoutWidth;
    
    NSString *timestamp = [[NSRelativeDateTimeFormatter new] localizedStringForDate:article.timestamp relativeToDate:NSDate.date];
    
    self.timeLabel.text = timestamp;
    self.timeLabel.accessibilityLabel = timestamp;
    
#if DEBUG_LAYOUT == 1
    self.titleLabel.backgroundColor = UIColor.greenColor;
    self.authorLabel.backgroundColor = UIColor.redColor;
    self.timeLabel.backgroundColor = UIColor.blueColor;
    self.secondaryTimeLabel.backgroundColor = UIColor.blueColor;
    
    self.backgroundColor = [UIColor grayColor];
    self.contentView.backgroundColor = [UIColor yellowColor];
#endif
    
}

- (void)configureTitle:(FeedVCType)feedType {
    
    if (feedType == FeedVCTypeNatural || feedType == FeedVCTypeAuthor) {

        self.titleLabel.text = self.article.articleTitle;
        return;

    }
    
    if ([self showImage] == NO) {
        
        self.titleLabel.text = self.article.articleTitle;
        return;
        
    }
    
    NSMutableParagraphStyle *para = [NSParagraphStyle defaultParagraphStyle].mutableCopy;
    para.lineSpacing = 24.f;
    
    NSDictionary *attributes = @{NSFontAttributeName: self.titleLabel.font,
                                 NSForegroundColorAttributeName: self.titleLabel.textColor,
                                 };
    
    NSMutableAttributedString *attrs = [[NSMutableAttributedString alloc] initWithString:formattedString(@"  %@", self.article.articleTitle) attributes:attributes];
    
    NSTextAttachment *attachment = [[NSTextAttachment alloc] init];
    
    Feed *feed = [MyFeedsManager feedForID:self.article.feedID];
    
    if (feed == nil || [self showImage] == NO) {
        
        self.titleLabel.attributedText = attrs;
        
        return;
    }
    
    NSString *url = [feed faviconURI];
    
    if (url != nil && [url isBlank] == NO) {
        NSString *key = formattedString(@"png-24-%@", url);
            
        [self _configureTitleFavicon:key attachment:attachment url:url];
    }
    
    // positive offsets push it up, negative push it down
    // this is similar to NSRect
    CGFloat fontSize = self.titleLabel.font.pointSize;
    CGFloat baseline = 17.f; // we compute our expected using this
    CGFloat expected = 7.f;  // from the above, so A:B :: C:D
    CGFloat yOffset = (baseline / fontSize) * expected * -1.f;
    
#if TARGET_OS_MACCATALYST
    attachment.bounds = CGRectMake(0, yOffset + 4.f, 16.f, 16.f);
#else
    attachment.bounds = CGRectMake(0, yOffset, 24, 24);
#endif
    
    NSMutableAttributedString *attachmentString = [NSAttributedString attributedStringWithAttachment:attachment].mutableCopy;
    
    [attachmentString appendAttributedString:attrs];
    
    self.titleLabel.attributedText = attachmentString;
    
}

- (void)_configureTitleFavicon:(NSString *)key
                    attachment:(NSTextAttachment *)attachment
                           url:(NSString *)url {
    
    if (self.faviconTask != nil) {

        // otherwise, cancel it and move on
        [self.faviconTask cancel];
        self.faviconTask = nil;
        
    }
    
    if (SharedPrefs.imageProxy) {
        
        url = [url pathForImageProxy:NO maxWidth:attachment.bounds.size.width quality:0.9f firstFrameForGIF:NO useImageProxy:YES sizePreference:ImageLoadingMediumRes];
        
    }
    
    if (url == nil) {
        attachment.bounds = CGRectZero;
        return;
    }
    
    self.faviconTask = [[SDWebImageManager sharedManager] loadImageWithURL:[NSURL URLWithString:url] options:kNilOptions progress:nil completed:^(UIImage * _Nullable image, NSData * _Nullable data, NSError * _Nullable error, SDImageCacheType cacheType, BOOL finished, NSURL * _Nullable imageURL) {
        
        if (error != nil) {
            
            NSLogDebug(@"Failed to fetch favicon at: %@", url);
            
            runOnMainQueueWithoutDeadlocking(^{
                
                attachment.bounds = CGRectZero;
                
                [self.titleLabel setNeedsDisplay];
                
            });
            
            return;
            
        }
        
        if ([image isKindOfClass:UIImage.class] == NO) {
            image = nil;
        }
        
        if (image != nil) {
            
            CGRect rect = attachment.bounds;
            rect.origin = CGPointZero;
            
            UIImage *rounded = [image sd_roundedCornerImageWithRadius:(3.f * UIScreen.mainScreen.scale) corners:UIRectCornerAllCorners borderWidth:0.f borderColor:nil];
            
            if (rounded != nil) {
                image = rounded;
            }
            
        }
        
        runOnMainQueueWithoutDeadlocking(^{
            if (image == nil) {
                attachment.bounds = CGRectZero;
            }
            else {
                attachment.image = image;
            }
            
            [self.titleLabel setNeedsDisplay];
        });
        
    }];
    
}

- (void)configureSummary {
    
    NSInteger previewLines = SharedPrefs.previewLines;
    
    if (previewLines == 0 || (self.article.summary && [[self.article.summary stringByStrippingHTML] length] == 0)) {
        self.summaryLabel.text = nil;
        self.summaryLabel.hidden = YES;
    }
    else {
        self.summaryLabel.hidden = NO;
        self.summaryLabel.numberOfLines = previewLines;
        self.summaryLabel.text = [self.article.summary stringByStrippingHTML];
    }
    
}

- (void)configureAuthorWithFeedType:(FeedVCType)feedType feed:(Feed *)feed {
    
    NSString *appendFormat = @" - %@";
    
    UILabel *authorLabel = self.authorLabel;
    
    authorLabel.hidden = NO;
    
    if (self.article.author) {
        
        if ([self.article.author isKindOfClass:NSString.class]) {
            
            if ([self.article.author isBlank] == NO) {
                
                authorLabel.text = [(self.article.author ?: @"") stringByStrippingHTML];
                
            }
            else {
                appendFormat = @"%@";
            }
            
        }
        else {
            
            authorLabel.text = [([self.article.author valueForKey:@"name"] ?: @"") stringByStrippingHTML];
            
        }
    }
    else {
        authorLabel.text = nil;
        appendFormat = @"%@";
    }
    
    if (feedType != FeedVCTypeNatural) {
        
        if (feed) {
            
            NSString *feedTitle = feed.displayTitle;
            
            if ([feedTitle isEqualToString:authorLabel.text] == NO) {
            
                authorLabel.text = [[(authorLabel.text ?: @"") stringByAppendingFormat:appendFormat, feedTitle] stringByStrippingWhitespace];

            }
            
        }
        else {
            
            authorLabel.text = nil;
            
        }
        
    }
    
}

- (void)configureCoverImage:(NSString *)url {
    
    if (url == nil) {
        return;
    }
    
    NSString *baseURL = [url copy];
    
    CGFloat maxWidth = self.coverImage.bounds.size.width;
    
    if (SharedPrefs.imageProxy == YES) {
        
        url = [url pathForImageProxy:NO maxWidth:maxWidth quality:0.9f firstFrameForGIF:NO useImageProxy:YES sizePreference:ImageLoadingMediumRes];
        
    }
    
    self.coverImage.contentMode = UIViewContentModeCenter|UIViewContentModeScaleAspectFill;
    
    [self.coverImage sd_setImageWithURL:[NSURL URLWithString:url] placeholderImage:[UIImage systemImageNamed:@"rectangle.on.rectangle.angled"] options:SDWebImageScaleDownLargeImages|SDWebImageRetryFailed completed:^(UIImage * _Nullable image, NSError * _Nullable error, SDImageCacheType cacheType, NSURL * _Nullable imageURL) {
        
        if (error != nil) {
            
            if (SharedPrefs.imageProxy == YES && [[error.userInfo valueForKey:SDWebImageErrorDownloadStatusCodeKey] integerValue] == 404) {
                // try the direct URL
                [self.coverImage sd_setImageWithURL:[NSURL URLWithString:baseURL] placeholderImage:[UIImage systemImageNamed:@"rectangle.on.rectangle.angled"] options:SDWebImageScaleDownLargeImages|SDWebImageRetryFailed completed:^(UIImage * _Nullable image, NSError * _Nullable error, SDImageCacheType cacheType, NSURL * _Nullable imageURL) {
                    
                    if (error != nil) {
                        
                        return;
                    }
                    
                    self.coverImage.contentMode = UIViewContentModeScaleAspectFill;
                    
                }];
                
            }
            
            return;
        }
        
        self.coverImage.contentMode = UIViewContentModeScaleAspectFill;
        
    }];
    
}

- (void)updateMarkerView {
    
    if (self.article.isBookmarked && self.feedType != FeedVCTypeBookmarks) {
        
        self.markerView.tintColor = UIColor.systemOrangeColor;
        self.markerView.image = [UIImage systemImageNamed:@"bookmark.fill"];
        
    }
    else if (self.article.isRead == NO) {
        self.markerView.tintColor = SharedPrefs.tintColor;
        self.markerView.image = [UIImage systemImageNamed:@"largecircle.fill.circle"];
    }
    else {
        self.markerView.tintColor = UIColor.secondaryLabelColor;
        self.markerView.image = [UIImage systemImageNamed:@"smallcircle.fill.circle"];
    }
    
}

@end

File: Yeti/Controllers/Feeds/ArticleCell.swift
//
//  ArticleCell.swift
//  Elytra
//
//  Created by Nikhil Nigade on 16/03/21.
//  Copyright © 2021 Dezine Zync Studios. All rights reserved.
//

import UIKit
import Models
import SDWebImage
import DBManager
import Combine

class ArticleCell: UITableViewCell {
    
    static let identifier = "articleCell"
    
    @IBOutlet var authorLabel: UILabel!
    @IBOutlet var coverImage: UIImageView!
    @IBOutlet var markerView: UIImageView!
    @IBOutlet var summaryLabel: UILabel!
    @IBOutlet var timeLabel: UILabel!
    @IBOutlet var titleLabel: UILabel!
    @IBOutlet var titleLabelWidthConstraint: NSLayoutConstraint!
    
    var cancellables: [AnyCancellable] = []
    
    weak var article: Article?
    var feedType: FeedType!
    
    fileprivate var isShowingCover: Bool {
        return !coverImage.isHidden
    }
    
    fileprivate var faviconTask: SDWebImageOperation?
    fileprivate var coverTask: SDWebImageOperation?
    
    override func awakeFromNib() {
        
        super.awakeFromNib()
        
        coverImage.layer.cornerRadius = 3
        coverImage.layer.cornerCurve = .continuous
        
        titleLabel.textColor = .label
        summaryLabel.textColor = .secondaryLabel
        authorLabel.textColor = .secondaryLabel
        timeLabel.textColor = .secondaryLabel
        
        selectedBackgroundView = UIView()
        
        #if targetEnvironment(macCatalyst)
        selectedBackgroundView?.backgroundColor = .systemFill
        selectedBackgroundView?.layer.cornerRadius = 6
        selectedBackgroundView?.layer.masksToBounds = true
        #else
        selectedBackgroundView?.backgroundColor = tintColor.withAlphaComponent(0.3)
        #endif
        
        separatorInset = .zero
        
        resetUI()
        
    }
    
    override func layoutSubviews() {
        
        super.layoutSubviews()
        
        guard let sbv = selectedBackgroundView,
              sbv.superview != nil,
              traitCollection.userInterfaceIdiom == .mac else {
            return
        }
        
        sbv.frame = bounds.insetBy(dx: 12, dy: 6)
        
    }
    
    override func prepareForReuse() {
        super.prepareForReuse()
        resetUI()
        
        for o in cancellables {
            o.cancel()
        }
        
        cancellables = []
        
    }
    
    func resetUI() {
        
        coverImage.sd_cancelCurrentImageLoad()
        
        for label in [titleLabel, summaryLabel, authorLabel, timeLabel] {
            label?.text = nil
        }
        
        coverImage.image = nil
        coverImage.isHidden = true
        
        markerView.image = nil
        
        faviconTask?.cancel()
        faviconTask = nil
        
        coverTask?.cancel()
        coverTask = nil
        
        semanticContentAttribute = .unspecified
        titleLabel.textAlignment = .left
        authorLabel.textAlignment = .left
        
    }
    
    override func tintColorDidChange() {
        
        super.tintColorDidChange()
        
        updateMarkerView()
        
        #if targetEnvironment(macCatalyst)
        selectedBackgroundView?.backgroundColor = .systemFill
        #else
        selectedBackgroundView?.backgroundColor = tintColor.withAlphaComponent(0.3)
        #endif
        
    }
    
    // MARK: - Configure
    var showImage: Bool {
        
        if SharedPrefs.articleCoverImages == false {
            return false
        }
        
        if SharedPrefs.imageBandwidth == ImageBandwidthOption.loadingNever {
            return false
        }
        
        else if SharedPrefs.imageBandwidth == ImageBandwidthOption.loadingOnlyWireless {
            return CheckWiFi()
        }
        
        return true
        
    }
    
    var isMicroblogPost: Bool {
        
        guard let article = article else {
            return false
        }
        
        if article.title?.isEmpty == true && article.content.count > 0 {
            
            if article.textFromContent != nil {
                return true
            }
            
            // find the first para
            let firstPara = article.content.reduce(nil, { (prev, cur) -> Content in
                
                if prev != nil && prev?.type == "paragraph" {
                    return prev!
                }
                
                return cur
                
            })
            
            return firstPara != nil
            
        }
        
        return false
        
    }
    
    func configure(_ article: Article, feedType: FeedType) {
        
        self.article = article
        self.feedType = feedType
        
        guard self.article != nil else {
            return
        }
        
        guard let feed = DBManager.shared.feedForID(article.feedID) else {
            return
        }
        
        configureTitle(feed: feed)
        
        if showImage == false {
            coverImage.isHidden = true
        }
        else {
            
            var coverImageURL = article.coverImage
            
            if coverImageURL == nil, article.content.count > 0 {
                
                // find the first image
                if let content = article.content.first(where: { (c) -> Bool in
                    
                    return c.type == "image"
                    
                }) {
                    article.coverImage = content.url
                    coverImageURL = article.coverImage
                }
                
            }
            
            configureCoverImage(url: coverImageURL)
        }
        
        configureSummary()
        configureAuthor(feed: feed)
        
        titleLabel.accessibilityValue = titleLabel.text?.replacingOccurrences(of: " | ", with: " by ")
        
        if (Paragraph.languageDirection(forText: article.title ?? titleLabel.text ?? article.summary ?? "") == NSLocale.LanguageDirection.rightToLeft) {
            
            semanticContentAttribute = .forceRightToLeft
            textLabel?.textAlignment = .right
            authorLabel.textAlignment = .right
            
        }
        
        let width = bounds.size.width - 48
        
        titleLabel.preferredMaxLayoutWidth = width - (isShowingCover ? 92 : 4) // 80 + 12
        
        titleLabelWidthConstraint.constant = titleLabel.preferredMaxLayoutWidth
        
        summaryLabel.preferredMaxLayoutWidth = width
        
        timeLabel.preferredMaxLayoutWidth = 92
        
        authorLabel.preferredMaxLayoutWidth = titleLabel.preferredMaxLayoutWidth - 24 - 12 - timeLabel.preferredMaxLayoutWidth
        
        let timestamp = RelativeDateTimeFormatter().localizedString(for: article.timestamp, relativeTo: Date())
        
        timeLabel.text = timestamp
        timeLabel.accessibilityLabel = timestamp
        
        updateMarkerView()
        
        article.$state
            .receive(on: DispatchQueue.main)
            .sink { [weak self] state in
                self?.updateMarkerView()
            }
            .store(in: &cancellables)
        
    }
    
    func configureTitle(feed: Feed) {
        
        guard let article = self.article else {
            return
        }
        
        if isMicroblogPost == true {
            
            titleLabel.text = article.textFromContent
            titleLabel.numberOfLines = max(3, SharedPrefs.previewLines)
            titleLabel.font = .preferredFont(forTextStyle: .body)
            
        }
        else {
            
            titleLabel.text = article.title
            titleLabel.font = .preferredFont(forTextStyle: .headline)
            
        }
        
        guard feedType != .natural && feedType != .author else {
            return
        }
        
        guard showImage == true else {
            return
        }
        
        let paraStyle: NSMutableParagraphStyle = NSMutableParagraphStyle()
        paraStyle.lineSpacing = 24
        
        let attributes: [NSAttributedString.Key: Any] = [
            .font: titleLabel.font!,
            .foregroundColor: titleLabel.textColor!
        ]
        
        let attrs = NSMutableAttributedString(string: " \(titleLabel.text ?? "")", attributes: attributes)
        
        var attachment = NSTextAttachment()
        
        configureTitleFavicon(attachment: attachment, url: feed.faviconURI, feed: feed)
        
        // positive offsets push it up, negative push it down
        // this is similar to NSRect
        let fontSize: Double = Double(titleLabel.font.pointSize)
        let baseLine: Double = 17 // compute our expected using this
        let expected: Double = 7 // from the above, A:B :: C:D
        var yOffset: Double = (baseLine / fontSize) * expected * -1
        
        yOffset += 6
        
        #if targetEnvironment(macCatalyst)
        attachment.bounds = CGRect(x: 0, y: yOffset, width: 16, height: 16)
        #else
        attachment.bounds = CGRect(x: 0, y: yOffset, width: 24, height: 24)
        #endif
        
        let attachmentString = NSMutableAttributedString(attachment: attachment)
        
        attachmentString.append(attrs)
        
        titleLabel.attributedText = attachmentString
        
    }
    
    func configureCoverImage(url: URL?) {
        
        guard let url = url else {
            coverImage.isHidden = true
            return
        }
        
        guard url.absoluteString.contains("core/emoji") == false else {
            coverImage.isHidden = true
            return
        }
        
        coverImage.isHidden = false
        
        var proxyURL = url
        
        if proxyURL.absoluteString.contains(".gif") {
            
            // only load covers for GIFs. Loading the full gif can
            // cause a lot of memory to be used for unpacking the
            // gif data and eventually crashing the app.
            
            let proxyPath = url.absoluteString.path(forImageProxy: false, maxWidth: coverImage.bounds.size.width, quality: 0.9, firstFrameForGIF: true, useImageProxy: true, sizePreference:  SharedPrefs.imageBandwidth, forWidget: false)
            
            if let aURL = URL(string: proxyPath) {
                proxyURL = aURL
            }
            
        }
        
        else if SharedPrefs.imageProxy == true {
            
            let proxyPath = url.absoluteString.path(forImageProxy: false, maxWidth: coverImage.bounds.size.width, quality: 0.9, firstFrameForGIF: false, useImageProxy: true, sizePreference:  SharedPrefs.imageBandwidth, forWidget: false)
            
            if let aURL = URL(string: proxyPath) {
                proxyURL = aURL
            }
             
        }
        
        coverImage.contentMode = .center
        
        DispatchQueue.global(qos: .userInteractive).async { [weak self, weak coverImage] in
            
            self?.coverTask = SDWebImageManager.shared.loadImage(with: proxyURL, options: [.scaleDownLargeImages, .retryFailed], context: nil, progress: nil, completed: { (image, data, error, cacheType, finished, imageURL) in
                
                guard finished == true else {
                    return
                }
                
                if let err = error,
                    proxyURL == url
                        && ((err as NSError).userInfo[SDWebImageErrorDownloadStatusCodeKey] as? Int) ?? 0 == 404 {
                    
                    #if DEBUG
                    print("Failed to download cover image with URL:", proxyURL)
                    #endif
                    
                    return
                    
                }
                
                guard let c = coverImage else {
                    return
                }
                
                DispatchQueue.main.async {
                    
                    c.image = image
                    c.contentMode = .scaleAspectFill
                    
                }
                
            })
            
        }
        
    }
    
    func configureSummary() {
        
        let previewLines = SharedPrefs.previewLines
        
        guard previewLines > 0 else {
            summaryLabel.isHidden = true
            summaryLabel.text = nil
            return
        }
        
        summaryLabel.isHidden = false
        summaryLabel.numberOfLines = previewLines
        summaryLabel.text = article?.summary
        
    }
    
    func configureAuthor(feed: Feed) {
        
        authorLabel.isHidden = false
        
        var format = " - %@"
        
        if let author = article?.author, author.isEmpty == false {
            
            authorLabel.text = author.stripHTML()
            
        }
        else {
            
            authorLabel.text = nil
            format = "%@"
            
        }
        
        if feedType != .natural {
            
            let feedTitle = feed.displayTitle
            
            if feedTitle != authorLabel.text {
                
                let text = (authorLabel.text ?? "")?.appendingFormat(format, feedTitle).trimmingCharacters(in: .whitespacesAndNewlines)
                
                authorLabel.text = text
                
            }
            
        }
        
    }
    
    func updateMarkerView() {
     
        guard let a = article else {
            return
        }
        
        if a.state == .Bookmarked && feedType != .bookmarks {
            
            markerView.tintColor = .systemOrange
            markerView.image = UIImage(systemName: "bookmark.fill")
            
        }
        else if a.state == .Unread {
            
            markerView.tintColor = SharedPrefs.tintColor
            markerView.image = UIImage(systemName: "largecircle.fill.circle")
            
        }
        else {
            
            markerView.tintColor = .secondaryLabel
            markerView.image = UIImage(systemName: "smallcircle.fill.circle")
            
        }
        
    }
    
}

extension ArticleCell {
    
    static func register(_ tableView: UITableView) {
        
        let nib = UINib(nibName: "ArticleCell", bundle: nil)
        tableView.register(nib, forCellReuseIdentifier: ArticleCell.identifier)
        
    }
    
    func configureTitleFavicon(attachment: NSTextAttachment, url: URL?, feed: Feed) {
        
        if faviconTask != nil {
            faviconTask?.cancel()
            faviconTask = nil
        }
        
        let config = UIImage.SymbolConfiguration(font: titleLabel.font!)
        
        attachment.image = UIImage(systemName: "square.dashed", withConfiguration: config)
        
        titleLabel.setNeedsDisplay()
        
        guard var url = url else {
            return
        }
        
        if SharedPrefs.imageProxy == true {
            let a = feed.faviconProxyURI(size: attachment.bounds.size.width) ?? url
            
            url = a
        }
        
        DispatchQueue.global(qos: .userInteractive).async { [weak self, weak attachment] in
            
            self?.faviconTask = SDWebImageManager.shared.loadImage(with: url, options: [], progress: nil, completed: { (image, data, error, cacheType, finished, imageURL) in
                
                guard let sself = self,
                      let a = attachment else {
                    return
                }
                
                guard error == nil else {
                    
                    #if DEBUG
                    print("Failed to fetch favicon at:", url, error!.localizedDescription)
                    #endif
                    
                    return
                    
                }
                
                guard let image = image else {
                    return
                }
                
                var bounds = a.bounds
                bounds.origin = .zero
                
                guard let rounded = image.sd_roundedCornerImage(withRadius: (3 * UIScreen.main.scale), corners: .allCorners, borderWidth: 0, borderColor: nil) else {
                    return
                }
                
                DispatchQueue.main.async {
                    
                    a.image = rounded
                    
                    sself.titleLabel.setNeedsDisplay()
                    
                }
                
            })
            
        }
        
    }
    
}
